File: native/src/query_fixture.zig
const std = @import("std");
const tide = @import("tide");
const mock_db = @import("mock_db.zig");
const pack = @import("pack.zig");
const hier = @import("hier.zig");

// Seam-B differential — the "zig-direct" side. Loads a VCD and, for every signal,
// samples pack.valueAt(db, id, tick) at (a stride of) its transition ticks — the
// SAME function the napi addon's getValueAt calls. Each result is dumped as a line
//
//     <id> <tick> <width> <x0hex> <x1hex>
//
// to the output file (arg 2). The JS harness (tests/differential.test.cjs) replays
// every (id, tick) through the production addon and asserts byte-equality. Any
// mutation introduced by the napi boundary (word packing, truncation, byte order,
// x/z loss) shows up as a diff — no oracle needed (METHODOLOGY §5).
//
//   zig build query-fixture -- <vcd> <out.txt> [--max-per-sig=N]
//
// x0/x1 are tide's storage bytes (LSB-first), hex lowercase. A trace whose ticks
// exceed 2^32 panics in mock_db.load (same as the addon) — the parent records the
// crash, exactly as the other suites do.

const MAX_PER_SIG_DEFAULT: usize = 400;

fn writePath(h: *const hier.Hierarchy, gpa: std.mem.Allocator, id: hier.NodeId, out: *std.ArrayList(u8)) !void {
    const node = h.nodes.items[id];
    if (node.parent) |p| {
        try writePath(h, gpa, p, out);
        try out.append(gpa, '.');
    }
    try out.appendSlice(gpa, node.name);
}

fn appendHex(out: *std.ArrayList(u8), gpa: std.mem.Allocator, bytes: []const u8) !void {
    const hexd = "0123456789abcdef";
    if (bytes.len == 0) {
        try out.append(gpa, '-'); // empty value plane
        return;
    }
    for (bytes) |b| {
        try out.append(gpa, hexd[b >> 4]);
        try out.append(gpa, hexd[b & 0xf]);
    }
}

pub fn main(init: std.process.Init) !void {
    const gpa = init.gpa;
    const io = init.io;

    var vcd_path: ?[]const u8 = null;
    var out_path: ?[]const u8 = null;
    var max_per_sig: usize = MAX_PER_SIG_DEFAULT;

    var positional: usize = 0;
    var it = init.minimal.args.iterate();
    _ = it.next(); // argv[0]
    while (it.next()) |a| {
        if (std.mem.startsWith(u8, a, "--max-per-sig=")) {
            max_per_sig = std.fmt.parseInt(usize, a["--max-per-sig=".len..], 10) catch max_per_sig;
        } else if (std.mem.startsWith(u8, a, "--")) {
            std.debug.print("unknown flag: {s}\n", .{a});
            return error.BadArg;
        } else switch (positional) {
            0 => {
                vcd_path = a;
                positional += 1;
            },
            1 => {
                out_path = a;
                positional += 1;
            },
            else => {},
        }
    }

    const vcd = vcd_path orelse {
        std.debug.print("usage: query-fixture <vcd> <out.txt> [--max-per-sig=N]\n", .{});
        return error.BadArg;
    };
    const out_file = out_path orelse {
        std.debug.print("usage: query-fixture <vcd> <out.txt> [--max-per-sig=N]\n", .{});
        return error.BadArg;
    };

    var loaded = try mock_db.load(gpa, vcd); // may panic on >u32 ticks (== addon)
    defer loaded.deinit();
    const h = &loaded.hierarchy;
    const end_t = loaded.end_t;

    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(gpa);
    var buf: [64]u8 = undefined;

    for (h.nodes.items) |node| {
        if (node.payload != .signal) continue;
        const id = node.payload.signal.handle;
        const q = loaded.db.query(id, 0, end_t) orelse continue;
        const len: usize = @intCast(q.len);
        if (len == 0) continue;

        // Stride so dense signals emit at most max_per_sig samples (always incl.
        // first + last). Deterministic — no RNG, no wall clock.
        const stride: usize = if (len <= max_per_sig) 1 else (len + max_per_sig - 1) / max_per_sig;
        var k: usize = 0;
        while (k < len) : (k += stride) {
            const tick = q.timestamps[k];
            const v = pack.valueAt(&loaded.db, id, tick) orelse continue;
            try out.appendSlice(gpa, std.fmt.bufPrint(&buf, "{d} {d} {d} ", .{ @intFromEnum(id), tick, v.width }) catch unreachable);
            try appendHex(&out, gpa, v.x0);
            try out.append(gpa, ' ');
            try appendHex(&out, gpa, v.x1);
            try out.append(gpa, '\n');
        }
        // Ensure the final transition is always covered.
        if (len > 1 and (len - 1) % stride != 0) {
            const tick = q.timestamps[len - 1];
            if (pack.valueAt(&loaded.db, id, tick)) |v| {
                try out.appendSlice(gpa, std.fmt.bufPrint(&buf, "{d} {d} {d} ", .{ @intFromEnum(id), tick, v.width }) catch unreachable);
                try appendHex(&out, gpa, v.x0);
                try out.append(gpa, ' ');
                try appendHex(&out, gpa, v.x1);
                try out.append(gpa, '\n');
            }
        }
    }

    try std.Io.Dir.cwd().writeFile(io, .{ .sub_path = out_file, .data = out.items });
    std.debug.print("query-fixture: wrote {d} bytes for {s}\n", .{ out.items.len, vcd });
}

File: native/src/bin/query_fixture.rs
//! The direct half of the marshalling differential.
//!
//! Loads a VCD and, for every signal in the hierarchy, samples the same value
//! function `getValueAt` calls at (a stride of) its transition ticks. Each result
//! is one line:
//!
//! ```text
//! <handle> <tick> <width> <min hex> <max hex> <z hex>
//! ```
//!
//! `tests/differential.test.cjs` replays every `(handle, tick)` through the
//! addon and asserts byte equality, so anything the Node-API boundary mutates —
//! word packing, truncation, byte order, a lost x or z — shows up as a diff with
//! no oracle needed. Planes are tide's storage bytes, least significant first, in
//! lowercase hex; an empty plane is `-`.
//!
//! ```text
//! query-fixture <vcd> <out.txt> [--max-per-sig=N]
//! ```

use std::fmt::Write as _;
use std::path::PathBuf;
use std::process::ExitCode;

use riptide::hierarchy::Node;
use riptide::pack::with_value_at;
use riptide::trace::{self, Loaded};
use tide_core::Samples;
use tide_core::metadata::{Id, Timestamp};

const MAX_PER_SIG_DEFAULT: usize = 400;
const USAGE: &str = "usage: query-fixture <vcd> <out.txt> [--max-per-sig=N]";

struct Args {
    vcd: PathBuf,
    out: PathBuf,
    max_per_sig: usize,
}

fn parse_args() -> Result<Args, String> {
    let mut positional = Vec::new();
    let mut max_per_sig = MAX_PER_SIG_DEFAULT;
    for arg in std::env::args().skip(1) {
        if let Some(value) = arg.strip_prefix("--max-per-sig=") {
            max_per_sig = value.parse().map_err(|_| format!("bad count: {value}"))?;
        } else if arg.starts_with("--") {
            return Err(format!("unknown flag: {arg}"));
        } else {
            positional.push(arg);
        }
    }
    if positional.len() < 2 {
        return Err(USAGE.to_owned());
    }
    Ok(Args {
        vcd: PathBuf::from(&positional[0]),
        out: PathBuf::from(&positional[1]),
        max_per_sig,
    })
}

fn main() -> ExitCode {
    let args = match parse_args() {
        Ok(args) => args,
        Err(message) => {
            eprintln!("{message}");
            return ExitCode::FAILURE;
        }
    };

    match run(&args) {
        Ok(bytes) => {
            println!(
                "query-fixture: wrote {bytes} bytes for {}",
                args.vcd.display()
            );
            ExitCode::SUCCESS
        }
        Err(message) => {
            eprintln!("query-fixture: {message}");
            ExitCode::FAILURE
        }
    }
}

fn run(args: &Args) -> Result<usize, String> {
    let loaded = trace::open(&args.vcd).map_err(|error| error.to_string())?;

    // Aliased declarations share a handle and so repeat here, which keeps the
    // dump a function of the file alone.
    let handles: Vec<Id> = loaded
        .hierarchy
        .nodes
        .iter()
        .filter_map(|node| match node {
            Node::Signal { handle, .. } if *handle != 0 => Some(Id(*handle)),
            _ => None,
        })
        .collect();

    let mut out = String::new();
    for id in handles {
        dump_signal(&loaded, id, args.max_per_sig, &mut out);
    }
    std::fs::write(&args.out, &out).map_err(|error| format!("{}: {error}", args.out.display()))?;
    Ok(out.len())
}

fn dump_signal(loaded: &Loaded, id: Id, max_per_sig: usize, out: &mut String) {
    let db = &loaded.trace.db;
    let Some(mut cursor) = db.samples(id, 0, loaded.end_t) else {
        return;
    };
    let Some(chunk) = cursor.next_chunk() else {
        return;
    };
    let times = chunk.times();

    // Stride so a dense signal emits at most `max_per_sig` samples, with the
    // first and last always among them. No randomness and no clock: the same
    // file always produces the same dump.
    let stride = if times.len() <= max_per_sig {
        1
    } else {
        times.len().div_ceil(max_per_sig)
    };

    let mut ticks: Vec<Timestamp> = times.iter().copied().step_by(stride).collect();
    if times.len() > 1 && (times.len() - 1) % stride != 0 {
        ticks.push(times[times.len() - 1]);
    }

    for tick in ticks {
        let line = with_value_at(db, id, tick, |sample, width| {
            let (min, max, z) = (hex(sample.min), hex(sample.max), hex(sample.z));
            format!("{} {tick} {width} {min} {max} {z}", id.0)
        });
        if let Some(line) = line {
            let _ = writeln!(out, "{line}");
        }
    }
}

/// tide's storage bytes as lowercase hex, least significant byte first. An empty
/// plane has no digits to print, so it reads as a dash.
fn hex(bytes: &[u8]) -> String {
    if bytes.is_empty() {
        return "-".to_owned();
    }
    let mut out = String::with_capacity(bytes.len() * 2);
    for byte in bytes {
        let _ = write!(out, "{byte:02x}");
    }
    out
}
